include/xtensor/xmath.hpp

```diff
@@ -231,6 +231,54 @@ namespace xt
         return detail::make_xfunction((functor_type)detail::clamp, std::forward<E1>(e1), std::forward<E2>(hi), std::forward<E3>(lo));
     }
 
+    namespace detail
+    {
+        template <typename T>
+        inline constexpr std::enable_if_t<std::is_signed<T>::value, T>
+        sign_impl(T x)
+        {
+            if (std::isnan(x))
+            {
+                return std::numeric_limits<T>::quiet_NaN();
+            }
+            return x == 0 ? copysign(T(0), x) : copysign(T(1), x);
```

Comment:
I was about to remove the constexpr, but I guess going with your solution will work as well.

The isnan is what numpy is doing. Not sure if required. 

You can decide! ;)